Add "Zoomed" mod

o-dasher opened the pull request:
This mod is based on a concept i made up myself.
The game reduces your "FOV (Field of View)" and you move your cursor across the playfield "Blind traveling" across the playfield.

https://user-images.githubusercontent.com/88356162/184784763-c50d9bec-b03f-41ea-a843-24099bf6fff4.mp4

A few notable changes included with this pr is the exposure of a mouse position in `ParallaxContainer` this is used because the Blind Travel mod changes the background parallax so it is more "immersible" when using auto mod. and the exposure of `OsuScreenStack`'s `parallaxContainer`.

I am pretty new to pull requests so any heads up will help and ~I don't have c# experience so feel free to correct me~ but personally everything feels sane.

Update

The mod now is called "Zoomed" 
The game zooms in the screen to your cursor, tracking it accordingly. The game scales the zoom even further the more you raise your combo.

Commit:
Creates osu! BlindTravel mod

Commit:
Create base test for blind travel mod

Commit:
Fix duplicated parallaxContainer check

Commit:
Move update parallaxContainer assert

Commit:
Remove unused dependency

peppy commented:
- The naming of this mod isn't great. It seems like it should draw from feeling like an FPS or tracking the cursor. "Blind" is too generic and doesn't really explain what the mod does at all.
- The transition after breaks seems weird and jarring. I'd probably make it much smoother or remove it altogether.
- The default tweening you have applied means there's a chance the screen won't track on fast jumps and the user will have to try and click off-screen circles. Is that the intention? If not, you might want to make it track more stringently. Also feel weird that the default value is also the minimum.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModBlindTravel.cs:
```diff
@@ -0,0 +1,227 @@
+
+        private const float default_scope_fov = 0.6f;
+
+        [SettingSource("Base FOV", "Adjust the base field of view")]
```
I'd probably call it zoom not FOV. FOV is generally used in terms of 3D, not 2D.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModBlindTravel.cs:
```diff
@@ -0,0 +1,227 @@
+        {
+            var scopePositionForDrawable = position + playfield.DrawSize / 2;
+
+            drawable.MoveTo(scopePositionForDrawable, cameraDelay);
```
You can't run transforms every frame. It won't be stable and it will add a lot of overhead.

Look into using the `Interpolation` helper class direction. Search for existing usages of `DampContinuously`.

o-dasher replied:
This and the other requests were fixed on the "overhaul" commit, thanks.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModBlindTravel.cs:
```diff
@@ -0,0 +1,227 @@
+            drawable.MoveTo(scopePositionForDrawable, cameraDelay);
+        }
+
+        private bool increasedVisibilityMode;
+
+        private Scheduler scheduler = new Scheduler();
+
+        private bool isTraveling;
+
```
Please move all your variables to the top of the class. Look at other classes for examples of how you should structure code.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModBlindTravel.cs:
```diff
@@ -0,0 +1,227 @@
+
+        private bool increasedVisibilityMode;
+
+        private Scheduler scheduler = new Scheduler();
```
This should not be required.

o-thiago replied:
I think i get all the other request, but how would i not need a scheduler for that? isn't it the simplest way

o-dasher replied:
Sorry for this double mention, messed up not switching accounts.
But how would the scheduler not be required like, it seems to be the simplest way to schedule something for the future isn't it?

peppy replied:
We don't really add a scheduler anywhere just for this kind of purpose.

You can just store a timestamp in this case, should work quite well.

mk56-spn commented:
playtested this a bit, as  i thought it has issues with bounding on tablets ( and apparently on mouse now that ive tried it) , im perosnally fine with offscreen circles i think its interesting learning them, but not being able to hit circles at all because you run up against the playfield border is no fun at all. i think it should remain as it right now but allowing you to extend out of your main tablet area  to the whole playfield size, this ofcourse poses an issue for full area players but i think thats a compromise that will have to be made

mk56-spn commented:
> And we can't forget about the mobile players who just want to tap the circles directly.

very good point but i think theyll have to make do with playfield scaling cant see how else one could work it out, kind of like how no scope will never really work on mobile

o-dasher commented:
Would "Cursor Focus" be a good naming?

Commit:
Using dampContinuously over transforming every frame

Flamiii commented:
Something like "Zoomed In" (similar to peppy's suggestion) would probably be better imo.

Walavouchey commented:
for reference, McOsu has an fps mod which doesn't have any delay in the screen tracking, which to me seems more reasonable <https://youtu.be/-o32vwQoyv4>

Commit:
Major overhaul

Commit:
Removes break increase visibility

This is a bad idea from the start.

Commit:
Fix cursor following drawables position calculation

Commit:
Set minimal baseZoom to 1.5f

Commit:
Rename "Blind Travel" to "Zoomed"

peppy commented:
Defaulting to zero delay would be my preference. It can still exist as a setting to give customisation.

o-dasher commented:
> Defaulting to zero delay would be my preference. It can still exist as a setting to give customisation.

I just did a Zero Delay local version and the screens feels too shaky, i am not sure but if you want i guess it is ok

Commit:
Adds possibility of zero cursor follow delay

I am not sure if this should be the default though.

peppy committed:
Fix abhorrent code style

peppy commented:
I've done a (mostly rider-automated) code style pass because your naming, styling and everything between was wrong.

Please check the contribution guidelines and read more existing code to make sure you get things right in the future.

I'm not reviewing this further because you've made changes outside the mod class (to `ParallaxContainer` and the likes). I haven't understood why this is required yet, but if it is, it should be in its own PR and tested separately.

Mod PRs should only touch the mod class.

Commit:
Remove out of pr scope changes and refactors

Commit:
Fix code style violations

o-dasher commented:
> I've done a (mostly rider-automated) code style pass because your naming, styling and everything between was wrong.
> 
> Please check the contribution guidelines and read more existing code to make sure you get things right in the future.
> 
> I'm not reviewing this further because you've made changes outside the mod class (to `ParallaxContainer` and the likes). I haven't understood why this is required yet, but if it is, it should be in its own PR and tested separately.
> 
> Mod PRs should only touch the mod class.

I am sorry for the naming of everything before, i should had paid more attemption for that and used a proper ide.

I modified the other classes because i was trying to do a increased parallax effect according with zoom, but tbh after testing it for a bit it kinda feels off anyway.

Commit:
Proper "Zoomed" mod test file name

Commit:
Update TestSceneOsuModZoomed.cs

Sorry.

Commit:
Simplifies code

peppy committed:
Merge branch 'master' into blind-travel-mod

peppy committed:
Fix code quality issues

peppy committed:
Move settings to top of file and reword / fix typos

peppy committed:
Remove pointless redirect properties

peppy commented:
I've pushed some initial fixes but I still have issues with this mod.

- Why is the combo adjust a bool rather than something closer to what `ModMuted` does (set the final combo count)? I'd argue it should *always* scale with combo since that feels like more fun, and just change the settings to more closely match what Muted is doing.
- I'm not sure about this giving silver S/SS. Probably need opinions from others.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,126 @@
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            currentZoom = InitialZoom.Value;
```
I don't think this is required to be here?

o-dasher replied:
Oh, yes it is indeed required, because at the start current zoom isn't being set and it differs from the base zoom which is just the base.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,126 @@
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            currentZoom = InitialZoom.Value;
+            gameplayClock = drawableRuleset.FrameStableClock;
```
I don't think this is required. You can use `Playfield.Clock` in the `Update` method directly, as the playfield is nested inside the frame stability container.

o-dasher replied:
Oh, i didn't knew about that thanks ig

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,126 @@
+        {
+            Debug.Assert(gameplayClock != null);
+
+            double currentScale = Interpolation.ValueAt(Math.Min(Math.Abs(gameplayClock.ElapsedFrameTime), apply_zoom_duration), playfield.Scale.X, currentZoom, 0, apply_zoom_duration, Easing.Out);
```
Is there a reason you didn't use `DampContinuously` here?

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,126 @@
+            else
+            {
+                playfield.Position = Interpolation.ValueAt(
+                    Math.Min(Math.Abs(gameplayClock.ElapsedFrameTime), MovementDelay.Value), playfield.Position, trackingPosition, 0, MovementDelay.Value, Easing.Out);
```
Same here, I think you can use `DampContinuously`. It may even avoid requiring the special handling above to prevent div-by-zero.

o-dasher replied:
i changed it to that because otherwise it would be required to do these if statements checking for some edge cases where when using playback with `dampContinuously` would cause an infinity number thus making the game crash, so i was doing 2 if statements for checking if the numbers aren't infinity, with the interpolation alone this isn't necessary

Commit:
Removes usage of gameplayclock

toadsworth-jr commented:
Maybe put it in the Fun category for now?
(as per a pull request I couldn't find)

mk56-spn commented:
Personally i wouldn't be opposed to this giving a silver S/SS as long as traceable and Blinds do so as well, considering those are far closer to HD and FL and more "traditional" mods than this

Commit:
Uses FinalZoomCombo instead of a bool

o-dasher commented:
> Maybe put it in the Fun category for now?
> (as per a pull request I couldn't find)

The mod is already on the fun category though (on this pull request of course).

Commit:
Removes unnecessary IApplicableToDrawableRuleset

Commit:
Removes unused variable

Commit:
Makes barrel roll mod incompatible with zoom

Barrel roll scales the playfield and rotates it, which makes cursor movement with a zoom applied pretty much unplayable.

Commit:
Fix code formatting

Commit:
Fix combo bindable

o-dasher commented:
The CI errors come from an outdated mod interface, should i merge master into this pr?

peppy committed:
Merge branch 'master' into blind-travel-mod

peppy committed:
Update `Description` to use `LocalisableString`

peppy commented:
There is zero chance this is doing what you intend it to do.

![JetBrains Rider 2022-08-24 at 05 17 42](https://user-images.githubusercontent.com/191335/186335017-c53fc91e-6921-4982-8222-3f4fbef74bc4.png)

Are you using an IDE?

Commit:
Fixes zoom combo scaling

o-dasher commented:
> There is zero chance this is doing what you intend it to do.
> 
> ![JetBrains Rider 2022-08-24 at 05 17 42](https://user-images.githubusercontent.com/191335/186335017-c53fc91e-6921-4982-8222-3f4fbef74bc4.png)
> 
> Are you using an IDE?

You are right, and yes i am using a ide (Visual Studio 2022), it didn't warn me about that though :(

peppy committed:
Merge branch 'master' into blind-travel-mod

peppy commented:
The combo zoom logic is completely broken in case you forgot to test this at all.

peppy committed:
Set better defaults

peppy committed:
Fix wrong english

peppy committed:
Remove pointless initial set logic

peppy committed:
Inline pointless extracted methods

peppy committed:
Fix more stuff

peppy committed:
Synchronise final combo specs with existing `ModMuted`

peppy committed:
Use `DampContinuously`

peppy committed:
Don't adjust rank colour

peppy commented:
I've rewritten stuff to be okay to my eye. Would be interested in some feedback on the values / how this actually feels.

Having the view not centered on the cursor is a bit weird. Maybe that should be a setting; maybe the default, not sure.

mk56-spn commented:
> I've rewritten stuff to be okay to my eye. Would be interested in some feedback on the values / how this actually feels.
> 
> Having the view not centered on the cursor is a bit weird. Maybe that should be a setting; maybe the default, not sure.

My complaint about the cursor being bound by the screen hasnt been adressed so many maps wont be possible on high delay for now, and by many i mean anything with any considerable jumps

from a organisation pov in terms of customization this is much better 

im still against 0 delay being the default since its so unfun compared to having some delay ( even having the flashlight level of delay as a base would be much more pleasant) but i guess thats not the end of the world since it is customizable 

thats all for me

mk56-spn commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,92 @@
+        {
+            Default = 100,
+            Value = 100,
+            MinValue = 0,
```
this should be 1 , 0 causes a crash ( not sure why, i assume it might be because it has yet to be updated with any combo information and is left with the combo zoomed value and the default value together? ) , and imo makes little sense regardless since if max zoom is being achieved at 0 combo then the setting the final and initial value to the same zoom does the job just fine

peppy replied:
Fixed.

mk56-spn commented on osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs:
```diff
@@ -0,0 +1,92 @@
+        public override ModType Type => ModType.Fun;
+        public override LocalisableString Description => "Big brother is watching your cursor.";
+        public override double ScoreMultiplier => 1;
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModBarrelRoll) };
```
playing with the Zoomed Mod with FL enabled currently hard crashes the game , might want to make them temporarily incompatible

peppy replied:
Fixed.

Feodor0090 commented:
Use 1.0 as start zoom and 2.0 as final.
![изображение](https://user-images.githubusercontent.com/53872073/187315318-2dfd6a59-c220-45a6-8f95-2a12516540f9.png)
The map starts for me without any hitobjects at all. Cursor is not visible.
![изображение](https://user-images.githubusercontent.com/53872073/187315106-f5ad9eca-3ade-45a6-97d0-8b4a8b0f99af.png)
You won't fail, hp just drains to zero and map plays to the end. Result screen won't appear, player will just freeze on last frame.
![изображение](https://user-images.githubusercontent.com/53872073/187315228-b0b7eb19-0352-45c9-80b9-145d6dd038d2.png)
Tested with 2 more beatmaps, the same.

Feodor0090 commented:
A proposal here: background should "parallax" too. Ideally, differently from playfield to save background<>playfield movement.

peppy commented:
> Use 1.0 as start zoom and 2.0 as final.

I agree with 1.0 starting (almost set this originally) but not sure about 2.0 final. It's pretty crazy.

peppy committed:
Adjust default values further based on feedback

peppy committed:
Avoid div-by-zero when combo increase is set to 0

peppy committed:
Use explicit mouse position to avoid feedback loop caused by sourcing from inside playfield

peppy commented:
I've adjusted things a bit further. It feels pretty good now. I think we should get this in the hands of players for further feedback.

peppy commented:
@smoogipoo requesting a very light review of this (just testing gameplay is fine) to make sure there's nothing you abhorrently disagree with.

smoogipoo commented:
Visual pass:

---

At 1.95x zoom (and 2x zoom), the playfield keeps scaling in to what I assume is the final scale even though I have delay and final zoom at combo set to 0. This happens with both autoplay on and off:

https://user-images.githubusercontent.com/1329837/187623315-39397572-18de-4e97-a2bf-eed6c26da18c.mp4

---

At 2x zoom, this is completely broken with autoplay:

### **DON'T WATCH IF EPILEPTIC**

https://user-images.githubusercontent.com/1329837/187622259-21a11487-9fe1-4e9f-b08e-9b0763c8e8bc.mp4

peppy commented:
So basically that's the same feedback loop I was looking to avoid, but it turns out to not be so simple to do so (the `RulesetInputManager` seems to somehow be *inside* the playfield...).

To explain, the mod is moving the playfield, which means even with a stationary cursor, the cursor's actual position will change based on the repositioning / scaling.

Will go back to the drawing board. May have to unapply the playfield's transform matrix.

o-dasher commented:
> A proposal here: background should "parallax" too. Ideally, differently from playfield to save background<>playfield movement.

I implemented this originally but doing that is out of scope of this pr because it would require to change some classes that are out of the mods scope, but this could be done in a separate pr.

Commit:
Fix minor typo

o-dasher commented:
> So basically that's the same feedback loop I was looking to avoid, but it turns out to not be so simple to do so (the `RulesetInputManager` seems to somehow be *inside* the playfield...?).
> 
> To explain, the mod is moving the playfield, which means even with a stationary cursor, the cursor's actual position will change based on the repositioning / scaling.
> 
> Will go back to the drawing board. May have to unapply the playfield's transform matrix.

That's one of the main challenges i had to fight while trying to make a setting for the cursor to be centered, it would be a really nice option but thinking on a way to actually make this implementable is hard

o-dasher commented:
> Visual pass:
> 
> ---
> 
> At 1.95x zoom (and 2x zoom), the playfield keeps scaling in to what I assume is the final scale even though I have delay and final zoom at combo set to 0. This happens with both autoplay on and off:

After looking a bit at the code i figured out on why is that so, although zoom is being set to the final combo if the Final Combo at is set to 0, it still interpolates the playfield from scale 1 to the final scale, this could be fixed by not using `DampContinuously` if the Final Combo is 0.

peppy commented:
The delay was intentional for the scale to smooth things off. I'll change that for the case of `FinalZoomCombo=0` though.

peppy committed:
Fix zoom scaling being applied when final combo zoom is set to zero

peppy commented:
The above mentioned feedback issue is due to this function:

https://github.com/ppy/osu/blob/27569e2ed5f0006d9066a05cf2f5d44cb703bc8a/osu.Game/Rulesets/UI/DrawableRuleset.cs#L319-L320

This is run by the `ReplayInputHandler` at the start of a frame, and has a one-frame delayed playfield position, causing the feedback loop.

https://user-images.githubusercontent.com/191335/188115772-5e13ba50-4ef5-4bf4-abf5-41da8f6096f0.mp4

Before attempting to fix this, I wanted to bring up something of concern, which is that this also means replays under this mod (or any other mod which adjusts the playfield) would still have replay data cursor positions stored **relative to the playfield**. I'm not sure if this is what we want?

It will change the path to fixing this issue, so I'll wait for opinions before continuing.

Feodor0090 commented:
If cursor is stored relative to playfield, we can "disable" mods (#19925) just by disabling them. 
If cursor is stored relative to screen/gameplay area/etc., mods like BR and this will require additional transforming of cursor position to place it where it was, for example. It also may be painful for external tools or server-side replays processor (if it exists or ever will be) because they will need to guess where the cursor exactly was.

It's logical to store cursor absolutely, where a player pointed it. But having it stored not in hitobjects' coordinate space may complicate some calculations now or in future, so maybe it's better to leave things as they is.

peppy committed:
Change playfield update call entry for mods to after children

peppy commented:
I've fixed the jumping issue by moving the update call to `UpdateAfterChildren`. I *think* this is more correctly in all cases (ensures any adjustments to the playfield apply to the next frame, so input / autoplay / replay can pick up on them on the correct frame)?

I can still get a bit of weird stuttering when the window isn't focused specifically, but that might be something to do with frame stability. Maybe this is okay for now?

o-dasher commented:
This seems to be the best solution for that in my oppinion, simply because this doesn't require any mod specific code and should work better for any future mod that ends up doing something similar.

smoogipoo commented:
> At 2x zoom, this is completely broken with autoplay:

This issue still hasn't been fixed.

peppy commented:
that's weird it worked correct for me, must be what i saw with frame stability. does it work if you set your frame limiter lower?

smoogipoo commented:
Doesn't seem to. I've tested all the way down to 60fps.

peppy committed:
Add test coverage of zero delay zoomed mod

peppy commented:
@smoogipoo can you take a look into the issue? i've added a test that repro's 100% (i think when i thought i fixed it yesterday i was testing on 1.8x instead of 2x, and apparently that's not enough to trigger it?).

Here's some pointer to the things which are interacting to cause the issue:

https://github.com/ppy/osu/blob/fbfab4aaa83c9497d683e1af60914448bc5bab4f/osu.Game.Rulesets.Osu/Mods/OsuModZoomed.cs#L79

https://github.com/ppy/osu/blob/19721a9bbbff59ca99f19989404885ed0e50d322/osu.Game.Rulesets.Osu/Replays/OsuFramedReplayInputHandler.cs#L26

https://github.com/ppy/osu/blob/ca05333bb9dbe4a6140ca3b291ae4335c3b587fa/osu.Game/Rulesets/UI/Playfield.cs#L246

It feels like there's a delay in `DrawInfo` matrix updates or something. But I've tried to debug and can't figure out why it's occurring yet.

Androidf13 commented:
@ekrctb

Androidf13 commented:
osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModZoomed.cs
